### DISCUSSION TITLE
Issue updating longhorn 1.0.2 -> 1.4.1

### DISCUSSION THREAD

**Discussion by SirSilly** [longhorn/longhorn#5775] [Q&A] [answered]
Hello,

I have an issue with Longhorn update to newest 1.4.1 version.

All started after updating RKE cluster from 1.19.x to 1.24.10 version.
Cluster upgrade finished successfully, however after cluster upgrade started facing issues with Longhorn: Volumes were failing to attach to pods. Basically in Longhorn UI it was showing that specific PVCs were attached to nodes (for example PVC-xxxx on Node-3), but when POD scheduled it said that it is unable to mount volume. Same was to volumes which weren't attached to any nodes, Longhorn UI recognized that volumes were assigned to pods, but pods stayed in "PENDING" mode and were unable to schedule because of that. 

![image (1)](https://user-images.githubusercontent.com/102787198/232602362-7ae36400-7221-411f-a2fc-62c80efea1d4.png)

![image2](https://user-images.githubusercontent.com/102787198/232602794-82d7f822-360d-47c5-b3c4-a9495cee32b1.png)

Later decided to give it a shot to upgrade Longhorn to newest version (from 1.0.2 to 1.4.1), because I thought that there might be some incompatibilities with old Longhorn version and newer kubernetes version. Update was done via kubectl command as previous Longhorn was not installed via rancher/helm and after that I noticed that longhorn-manager did not start with error message:

`level=fatal msg="Error starting manager: admission webhook \"validator.longhorn.io\" denied the request: failed to set the setting system-managed-components-node-selector with invalid value : cannot modify node selector for managed components setting before all volumes are detached"`

Judging from the error message it seems that it cannot do any changes because not all volumes are detached, however all pods are scaled down to 0, so no pods are actually using the volumes.
I tried changing nodeIDs and currentOwnerID in Volume CRD, but it kept being changed back.

As well now, Longhorn UI does not load, but gives a blank dashboard where none of PVCs nor Nodes volume information are visible.

Any ideas how that could be resolved? How could I manually detach the volumes or somehow finish the upgrade?

Please let me know if you need extra information.

Thank you.

**Comment by c3y1huang** [comment]
> Later decided to give it a shot to upgrade Longhorn to newest version (from 1.0.2 to 1.4.1), because I thought that there might be some incompatibilities with old Longhorn version and newer kubernetes version. 

Did you upgrade your version incrementally, moving from 1.0.2 to 1.1.x, then to 1.2.x, and so on, or did you upgrade directly from 1.0.2 to 1.4.1?

>  How could I manually detach the volumes

```
kubectl -n longhorn-system patch volume <name-of-volume>  -n longhorn-system --type json -p '[{"op": "remove", "path": "/spec/nodeID"}]'
```

**Reply by SirSilly** [reply]
> Did you upgrade your version incrementally, moving from 1.0.2 to 1.1.x, then to 1.2.x, and so on, or did you upgrade directly from 1.0.2 to 1.4.1?

Upgrade was directly from 1.0.2 to 1.4.1, not incrementally.


> ```
> kubectl -n longhorn-system patch volume <name-of-volume>  -n longhorn-system --type json -p '[{"op": "remove", "path": "/spec/nodeID"}]'
> ```

Unfortunately after using this command it still shows as "attached"

Maybe it would be simpler solution just to reinstall longhorn completely? But if doing so, is there a possibility to get PVCs stored somewhere (even like locally in my PC) that after uninstalling longhorn they still would be possible to recover back to Cluster?

Thank you.

**Reply by c3y1huang** [reply, marked as answer]
> Upgrade was directly from 1.0.2 to 1.4.1, not incrementally.

Longhorn only supports [incremental upgrades](https://longhorn.io/docs/1.4.1/deploy/upgrade/longhorn-manager/#upgrading-from-v14x--vhahahugoshortcode-s0-hbhb-or-v13x).

> Unfortunately after using this command it still shows as "attached"

I see, this is ineffective because the longhorn-manager is down.

> Maybe it would be simpler solution just to reinstall longhorn completely? But if doing so, is there a possibility to get PVCs stored somewhere (even like locally in my PC) that after uninstalling longhorn they still would be possible to recover back to Cluster?

To restore the data, simply applying the same PVC CRs won't work. Instead, you could try to [recover the data from the replicas](https://longhorn.io/docs/1.4.1/advanced-resources/data-recovery/export-from-replica/) , then save and transfer them to a new Longhorn system.

**Reply by SirSilly** [reply]
Thank you for your answer and provided information.
I have managed to get data from PVCs, currently left running on openEBS some system with copied data, but in the future will remove completely longhorn and install fresh.

Thank you.
